### ISSUE TITLE
storage: gc queue processing times out during txn push/intent resolution

### ISSUE THREAD

**Issue by matbhuvi** [cockroachdb/cockroach#34082]
I have a 5 node cluster. Seeing a pattern where GC Queue and compaction queue is not coming down. Wondering whether these are impacting our latencies. I had issue with raft log queue sitting at 6s as well. But running `cockroach debug compact <data folder>` helped to resolve that. That did not help to bring down GC & Compaction queue though.

```
Build Tag:    v2.1.0
Build Time:   2018/10/30 12:32:34
Distribution: CCL
Platform:     linux amd64 (x86_64-unknown-linux-gnu)
Go Version:   go1.10.3
C Compiler:   gcc 6.3.0
Build SHA-1:  f0dd1dc2d2500155c8b36330ae3d6932c13c1027
Build Type:   release
```

<img width="1005" alt="GC" src="https://user-images.githubusercontent.com/23704258/51309605-d261ef80-1a6a-11e9-8fe4-5880a548d652.png">

<img width="1013" alt="GC" src="https://user-images.githubusercontent.com/23704258/51309634-e0b00b80-1a6a-11e9-9ef9-bf3dfbce504c.png">

Notable log i can see is

```
E190117 08:07:18.369770 10766946 storage/queue.go:793 [n4,gc,s4,r5909/4:/Table/66/9/"100{1977…-2020…}] failed to push during intent resolution: context deadline exceeded W190117 08:07:25.086071 642 server/node.go:886 [n4,summaries] health alerts detected: {Alerts:[{StoreID:4 Category:METRICS Description:queue.gc.process.failure Value:1}]}
```

Attached complete log from a node:
[cockroach.log.gz](https://github.com/cockroachdb/cockroach/files/2768011/cockroach.log.gz)

**Comment by tbg** [comment]
Hi @matbhuvi, the compaction queue paces its work, so it's often full like that. The GC queue, on the other hand, shouldn't have this behavior - it should be empty most of the time.

1. Could you upgrade the cluster to v2.1.3? There's no relevant bug fix that I know of, but it can't hurt. The upgrade is as easy as switching out the binaries and restarting each node.
2. The health alert you've pointed out is relevant. It's connected to this message:
    > E190117 09:33:29.859883 24449841 storage/queue.go:793  [n4,gc,s4,r5617/3:/Table/66/10/"10016{425…-873…}] failed to push during intent resolution: context deadline exceeded
4. the operation that causes this log message runs during GC, and it timing out suggests that there are lots of intents or that there are some ranges in your cluster that are unavailable.

What workload are you running against this cluster, in particular qps and roughly what kind of queries you're issuing.

Could you post a printout of `8080/#/reports/range/5797` as well as the `:8080/#/reports/problemranges` page (if there are many items on it, print the range status for some of each category as well please).

I'd also like a `./cockroach debug zip` run against n4, please.

Finally, `:8080/#/debug/enqueue_range` with the following: 
![image](https://user-images.githubusercontent.com/5076964/51313231-5e651e80-1a4d-11e9-954f-1016c6d4687c.png)

**Comment by matbhuvi** [comment]
Thanks for taking this up @tbg 

The workload is simple insert into couple of tables (each with 10, 15 columns respectively) at 500 tps.

I have attached the reports that you have asked for.

[Enqueue_Range.pdf](https://github.com/cockroachdb/cockroach/files/2769005/Enqueue_Range.pdf)
[ProblemRanges.pdf](https://github.com/cockroachdb/cockroach/files/2769011/ProblemRanges.pdf)
[Range_report.pdf](https://github.com/cockroachdb/cockroach/files/2769012/Range_report.pdf)

The debug zip file is coming to 150+ MB. Can i filter out and upload anything specific?

I'll update to 2.1.3 and see whether there is a difference as well.

**Comment by tbg** [comment]
Thank you @matbhuvi. From Enqueue_Range I can see that there were 221621 provisional writes in this range. That's a very large number and matches the MVCC Intent Bytes/Count in ProblemRanges.pdf.

I'm also concerned by the ProblemRanges page, which indicates timeouts for reaching all of the nodes (confusingly, the other pages work, so this is perhaps an unrelated problem).

I'd like to see your workload, please. Something is creating these intents (i.e. transactional writes that don't clean themselves up) at a high rate. Could you share your schema and the transactions you run?

**Comment by tbg** [comment]
PS if you'd like to share any of that privately, tobias@cockroachlabs.com will work.

**Comment by matbhuvi** [comment]
I have shared you the schema and sample queries via email. As i said, it is just two simple tables and insert into them. We did try with high number of indexes in each table which proved to be a bad choice. We are re-working the model based on performance.

**Comment by tbg** [comment]
Thank you @matbhuvi. I just took a look, seems that r5797 is part of `guest_order`. Am I understanding correctly from your email that you're mostly inserting single rows (i.e. no bulk inserts)?

Also, if this is a testing environment, can you check whether the GC queue pending count goes down over a few hours when no inserts are run against the cluster?

Third, can you run a `COUNT(*)` over `guest_order` and see how that affects the range status of 5797 and in particular its "MVCC Intent Bytes/Count" field?

I'm going to loop in my colleague @andreimatei to help out.

**Comment by tbg** [comment]
Ah, a correction to my earlier message it's an *index* on `guest_order`, not the primary index itself. Unfortunately I'm not quite sure which one right off the bat. This means that for the `COUNT(*)` to work the index would have to be forced, I think this works using `table@index_name` instead of `table`.

**Comment by tbg** [comment]
You can find out the index via `select * from crdb_internal.table_indexes;` and searching for descriptor_id=66 and index_id 12.

**Comment by andreimatei** [comment]
Hi @matbhuvi! We'd certainly like to understand what happened / is happening to your cluster.

So, we know that there are a ton of intents everywhere: I see the log messages that you've indicated over and over again, on many ranges.
```
failed to push during intent resolution: context deadline exceeded
```
An "intent" is a provisional value written as part of a transaction. When the transaction commits, the intent is turned into a committed value asynchronously. So assuming transactions are short-lived (and you seem to be doing simple queries), intents are supposed to be short-lived too. Having many of the laying around is bad. We need to figure out how they're being left behind.

Then, beyond the fact that there's apparently tons of intents everywhere, we also see that the GC queue is having trouble clearing them - again the "failed to push... deadline exceeded". I believe the deadline in question is 1 minute, used by the GC Queue for all its operations. The "push" refers to the GC Queue trying to determine if the transactions that laid down the intents it's trying to clean up have committed or not. And this operation times out... Which I think means that the GC Queue cannot make progress; it will retry to do the same thing over and over. I'm looking at the code and it looks to me like the code that deals with these intents is making some attempts to deal with situations where it has very many intents to deal with - it tries to batch them so that some forward progress can be made - but not for the particular step that's timing out here. I think having intents belonging to many different transactions could potentially cause the operation in question to take a long time. This might be something we should improve for the robustness of this code.
I was thinking that we'd get more clues about the cleanup work done that needs to be done by the GC Queue from that `enqueue_range` report you've generated from us. Unfortunately, there's very little there except the number of intents. In particular, I would have liked to see what this line says about how many transactions are involved:
https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/intent_resolver.go#L593
The best I can tell after trolling the code is that that line didn't fire because we early-returned above it in that function because someone else was in the middle of pushing the same intents. But that seems like a big coincidence... I'm not sure what to make of it.

Now, back to the original mystery - why are there so many intents anywhere? I'm not sure about the best way to get clues about that. Whether or not all the intents belong to a single transactions or to millions of them would be useful information, but I'm not sure how to get it.
One thing that might help is if you told us what queries are currently running.
If you could run `show cluster queries` and `show cluster sessions` from a SQL prompt and give us the results, it might help.
Another thing that might help is the `./cockroach debug zip` on node 4. We didn't get that so far, did we? Would it be possible for you to put it on google drive of Dropbox for a little while until we download it? That zip has a snapshot of goroutine dumps so we can see what the server is working on.
@tbg's suggestion for trying to do a `select count(*)` from that index is also good. Such a query is supposed to clean up all the intents it runs into, so it'd be interesting to see if it succeeds in doing that.

Or better yet, is this a production cluster? If not, do you think it might be possible for me and my Tobi to poke around it for a few hours sometime? In this kind of open-ended investigation, having direct access is _much_ quicker. You cluster's situation is interesting and scary, so we'd do our best to understand it.

**Comment by andreimatei** [comment]
One interesting thing I see in the logs from node 4 is errors from the split queue:
```
E190118 01:17:22.238409 6378629 storage/queue.go:940  [n4,split] 1 replicas failing with "could not find valid split key"
```
This means that some range (or perhaps many of them) should be split because they're too large, but we're failing to split them because they consist of a single row. This can mean that the row is simply very large (over 64MB I think) or perhaps the row was updated a lot and so it has a lot of historical versions. Does your workload involve either large rows or repeated updates to the same row? I don't see `updates` in the show queries.

Then, interestingly, I see these starting at 03:42:09 GMT. 
```
W190118 03:47:52.579543 15633 storage/replica_backpressure.go:135  [n4,s4,r7/11:/Table/{SystemCon…-11}] applying backpressure to limit range growth on batch [txn: a33ba6c8], BeginTransaction [/Table/SystemConfigSpan/Start,/Min), Put [/Table/3/1/66/2/1,/Min)
```
This seems oddly related, although the other errors are present in the logs before this one too. This message means that one of the ranges which holds the table catalogue has gotten to be too big and we're refusing writes to it. Writes to that system catalogue are done by schema changes, in this case it looks like a schema change on table 66 which is our friend `guest_order`. I'm guessing this is the `drop index` that you've attempted and that got stuck.
I think this is an opportunity for improvement - we should have just failed that schema change rather then keep it retrying forever which I think is what we do. How this particular range got big is beyond me. I can't see how I'll open an issue for that. You say that some schema changes got "cancelled" - can you detail what you means? Can you also give us the output of `show jobs` please?

Anyway, all these clues I think tell a story about why the GC queue is now going down - I think the GC queue doesn't operate on non-split ranges, but the split queue is probably stuck on a couple of un-splittable ranges (and some of them maybe are un-splittable exactly because GC isn't running?).

There's also the fact that the `debug.zip` doesn't have anything about node 6. Is that node running? Does the list of nodes in the Admin UI show it as green and happy? Maybe there's something wrong with it and perhaps that explains why the Problem Ranges report is not loading.

All this still doesn't say anything about how your cluster got to have tons of intents in the first place. You don't seem to have long-running complex transactions. I've checked all the stack traces from the debug.zip and there didn't seem to be any transactions running at all at that moment. Perhaps it was dropping an index somehow left a ton of intents everywhere. A very interesting thing to me would be to figure out what intents there still are in the cluster and try to understand where they came from. Could you share your cluster's data with us? That would allow us to do forensics on it. If you can, I'd be interested in you simply making copies of all the store directories and giving them to us.

Also, if possible, it'd be great if you not touch the cluster at all for a while while we figure out next steps. Is this acceptable? Don't put any load on it, don't restart any nodes, etc. Btw, have you restarted any nodes recently?
I think the ball is in our court to figure out the next investigation steps.
To summarize:
- if you can, don't touch the cluster
- if you can, send us copies of the store directories
- send us `show jobs`


A Skype call next week might also be very good, but we need to figure out exactly what to ask you for.

**Comment by andreimatei** [comment]
We have some ideas about some next steps, but it's going to take me a bit to craft some command lines that you can run to explore some low-level data in the store. If you could give some stores to us, that'd still be better.

**Comment by matbhuvi** [comment]
>  You say that some schema changes got "cancelled" - can you detail what you means? Can you also give us the output of show jobs please?

I have already shared that in the google drive.

> There's also the fact that the debug.zip doesn't have anything about node 6. Is that node running? Does the list of nodes in the Admin UI show it as green and happy? Maybe there's something wrong with it and perhaps that explains why the Problem Ranges report is not loading.

It has beed decommissioned. I have shared the cluster overview page for further details.

> Btw, have you restarted any nodes recently?

Yes. As part of 2.1.3 upgrade that we did yesterday based on @tbg 's recommendation.

> if you can, send us copies of the store directories

It is 1.2 TB. I need to figure out how to share that.

> Before you do anything else, please generate another debug.zip and give it to us.

I have uploaded the latest one.

Also uploaded the debug keys you have asked for and crdb_internal queries.

**Comment by andreimatei** [comment]
Thanks @matbhuvi!
I don't see the `rangeid:7` data dump in that Google Drive dir. Have I missed it?
And btw I gave you incomplete instructions for that. What I want is: `./cockroach debug keys --from='rangeid:7' --to='rangeid:8' --values <store path>`.

Thanks for bearing with us!

**Comment by matbhuvi** [comment]
> I don't see the rangeid:7 data dump in that Google Drive dir. Have I missed it?
And btw I gave you incomplete instructions for that. What I want is: ./cockroach debug keys --from='rangeid:7' --to='rangeid:8' --values <store path>.

Sorry for the delay. I have uploaded debug keys from rangeid:7 to rangeid:8

**Comment by andreimatei** [comment]
I'm sorry Mathan, the command I gave you for dumping range 7 was not correct. Ugh. The command I gave you dumped some specific "range local" keys, not generally the contents of the range.
The right command is:
```
./cockroach debug keys cockroach-data/ --from='raw:\x88' --to='raw:\x93' --values
```

**Comment by andreimatei** [comment]
So looking at the data from the system config range, I see 152k versions of the tab

**Comment by andreimatei** [comment]
So, looking at the data from the SystemConfigRange, I see 152k versions of the guest_order table descriptor. That's... not good. I've decoded the dumped descriptors; looking at a few consecutive versions in there, they differ by (alternatively):
```
>   lease: <
>     node_id: 1
>     expiration_time: 1547783229586797876
>   >
------
<   lease: <
<     node_id: 1
<     expiration_time: 1547783229828965854
<   >
```

So, it would appear that a schema change keeps retrying and, every time, it acquires and releases a "schema change lease" on the table. These piled up over time and caused this special, unsplittable, SystemConfigSpan to grow to the point where it doesn't allow writes any more because it's gotten too big. So, at this point, I believe no schema change can succeed any more any generally the cluster will have major problems because of this.
All these old descriptor versions are supposed to be cleaned up by the GC queue, but as we've seen before the GC queue is probably stuck trying to deal with some other ranges and fails because those other ranges have too many intents (in particular, intents from too many different transactions).

So now let's turn our attention back to the intents. What I can see from the index dumps you've sent is that there's a ton of intents, every single one belonging to its own transaction. Interestingly, all these intents seem to have been laid down in a very narrow time window:
Table 66 index 6: 1546523760 to 1546524622 (Thursday, January 3, 2019 1:56:00 PM GMT to 2:10:22 PM)
Table 66 index 8: 1546523760 to 1546524622 (Thursday, January 3, 2019 1:56:00 PM GMT to 2:10:22 PM)
Table 66 index 9: 1546525296 to 1546527322 (Thursday, January 3, 2019 2:21:36 PM GMT to 2:55:22 PM)

Interestingly, the records for index 6 and index 8 match (in time and transactions). And the index 8 comes soon after, but completely non-overlapping. I wonder all this weird matching is because I've only asked you for the first 50000 intents from each index; perhaps the matching would be perfect if I were looking at all the intents.
I don't know how to figure out why these intents were left over. We have no logs for January 3rd any more. Perhaps you can remember if there was a particular event back then - something like a node crashing? We expect to leave intents laying around when a node crashes, although it's unusual to leave many... It would mean that somehow a lot a transactions were in flight at the time of the crash and, according to our timing info, some of these transactions would have had to be stuck for like 30m before the crash... Perhaps doing `select * from system.eventlog` to see when nodes restarted will show something for Jan 3rd.

In any case, my fear was that we have some bug causing intents to more routinely be left around, and I guess there's no evidence of that.

The main improvements to be made, I think, coming out of this investigation are:
a) We really need to stop thrashing the SystemConfigSpan with table descriptors. I think the schema changer should stop writing lease-on/lease-off updates all the time and I really think that all the chatter related to schema changes should move out of the SystemConfigSpan. This key space is too fragile because it's gossipped and it can't be split. I've advocated for this before; I think everything about mutations should move out of the table descriptor, and definitely this schema change lease record I think.
@vivekmenezes I think this part falls on you. I haven't created a new issue for it since there's a lot of context in this thread; I'll leave that to you.

b) The GC queue should not get stuck on the `failed to push during intent resolution: context deadline exceeded` error. More generally, the `intentResolver` should be smarter around batching the transactions to push: `maybePushTransactions` should batch its work, like it does with the intent cleanup part. @ajwerner I think you're currently poking around that code. You want this one? Also feel free to create a new issue.

c) We need to rethink write backpressure on the SystemConfigSpan. Should there be backpressure given that the range can't be split? Should there be some other alert? Will sleep on it.
Related: #34210

Finally, brother @matbhuvi , what's the current status of this cluster? I assume not good. Are you interested in restoring it to a working condition?
If so, I'd start by trying to force a cleanup of the intents:
```
select count(1) from guest_order_guest_id_card_otep_token_idx;
select count(1) from guest_order@guest_order_phone_number_idx;
select count(1) from guest_order@guest_order_updated_date_idx; -- this one might fail cause the index is in the process of being dropped
```
And I'd force a GC run on the `SystemConfigSpan` range:
Go to `/#/debug/enqueue_range` in the Admin UI, select `Queue: gc`, fill in
`RangeID: 7` and check `SkipShouldQueue`. Let me know how it went. And thanks a lot for your time.

**Comment by vivekmenezes** [comment]
@andreimatei created https://github.com/cockroachdb/cockroach/issues/34211 to move the schema change lease.

**Comment by matbhuvi** [comment]
> Perhaps doing select * from system.eventlog to see when nodes restarted will show something for Jan 3rd.

I don't see any event between 2018-12-18 and 2019-01-09. I have shared the event logs.

> If so, I'd start by trying to force a cleanup of the intents:
Except the 2nd one, the others are dropped now. 


>Go to /#/debug/enqueue_range in the Admin UI, select Queue: gc, fill in
RangeID: 7 and check SkipShouldQueue. Let me know how it went. 

```
2019-01-24 05:37:55 | [n3] running gc.process
2019-01-24 05:37:55 | [n3] skipping replica; low score queue=false with 0.74/fuzz(1.00)=0.74=valScaleScore(0.80)*deadFrac(0.93)+intentScore(0.00) likely last GC: 24h10m26.126195557s ago, 279 KiB non-live, curr. age 19 GiB*s, min exp. reduction: 0 B*s
```

Triggered it. GC Queue has gone down from 960 to 300 in 2hrs. Will need to see whether it completely goes out.

```
[n3,gc,s3,r5500/3:/Table/66/8/"10016{037…-300…}] failed to push during intent resolution: context deadline exceeded
```

> what's the current status of this cluster? 

I have spun up a new cluster for my team to continue their POC. This cluster is completely dedicated to understand and resolve this issue.

**Comment by tbg** [comment]
@andreimatei could it be that the intents were laid down by an index backfill? You say you're seeing adjacent writes belonging to different txns, but isn't that what would happen during a chunked backfill of an index?

Say we have PKs

```
A/C
B/E
C/B
D/D
E/A
```

and an index on the second component gets created, with a batch (=txn) size of 2. Then we'd see the writes in three txns:

```
C
E
```

```
B
D
```

```
A
```

and their intents on the indexes would be all interspersed.

Is that perhaps what you're seeing?

@matbhuvi
> Will this error goes away once GC queue empties?

Yep, this should go away. You can tell be the `gc` part in the front that this comes from the GC queue. It's related to there being more work to be done than it can finish in a minute (it should continually make progress though).

> Triggered it. GC Queue has gone down from 960 to 300 in 2hrs. Will need to see whether it completely goes out.

That output tells me that the GC queue didn't do anything because there's very little to be gc'ed. I assume this means that the GC happened before already, though the message also indicates that it's been ~24h at which point I thought we still saw the problems. @andreimatei did you give out the right RangeID?

**Comment by andreimatei** [comment]
> You say you're seeing adjacent writes belonging to different txns, but isn't that what would happen during a chunked backfill of an index?

I see no reason to believe that these intents had been left over by index backfills, and so I'm inclined to not believe it. The chunk size for an index backfill is 100; I've only seen txn have two intents. Also the intents from the txns are spread across two indexes, so a backfill wouldn't do that.
I'm more inclined to believe that something had gone wrong with the cluster on January 3rd while normal insert traffic was ongoing, and a lot of transactions that laid down intents were unable to commit. Or... who knows; I've kinda lost hope that we can figure out what happened there, and also it seems to have been an event localized in time, so I think everybody's time is better spent elsewhere. There's enough of meat that came off of this investigation already to feed us for a while.

And there might be yet more to come as we try to bring this cluster back to life, which I think is worth doing if @matbhuvi will bear with us - since so many things have already gone wrong here, maybe we discover more actionable items as the cluster recovers.


> > Will this error goes away once GC queue empties?
> > [n3,gc,s3,r5500/3:/Table/66/8/"10016{037…-300…}] failed to push during intent resolution: context deadline exceeded

> Yep, this should go away. You can tell by the gc part in the front that this comes from the GC queue. It's related to there being more work to be done than it can finish in a minute (it should continually make progress though).

What's interesting here is that there's still intents on this index. Index 8 is `guest_order_guest_id_card_otep_token_idx`, and you've said that this index is dropped now. But the act of dropping it was supposed to clear up all the intents. So... how dropped is it?
Would you mind doing another `select * from crdb_internal.table_indexes`
and also giving the output of `show jobs`?


> > Triggered it. GC Queue has gone down from 960 to 300 in 2hrs. Will need to see whether it completely goes out.

> That output tells me that the GC queue didn't do anything because there's very little to be gc'ed. I assume this means that the GC happened before already, though the message also indicates that it's been ~24h at which point I thought we still saw the problems. @andreimatei did you give out the right RangeID?

I sure think so (pretty sure it's range 7). Would you mind giving us the output of `select * from crdb_internal.ranges limit 20;`? That'll answer definitively what range id we want for the `SystemConfigSpan`.

Since this cluster is our playground now, would it be possible for us to get ssh access to it? If that's OK, and if you need to whitelist some IP for ssh, you can whitelist `35.235.98.54`. If you give us access, you don't need to do any of the things I've asked about above. We can watch the cluster ourselves.

**Comment by andreimatei** [comment]
> My bad. The index seems to be present. But, getting this error when selecting from it.
> pq: relation "guest_order_guest_id_card_otep_token_idx" does not exist

I gave you wrong instructions. I wanted to say `select count(1) from guest_order@guest_order_guest_id_card_otep_token_idx`, to clear intents from that index.

> > Would you mind doing another select * from crdb_internal.table_indexes
and also giving the output of show jobs?
> > Would you mind giving us the output of select * from crdb_internal.ranges limit 20;? That'll answer definitively what range id we want for the SystemConfigSpan.

> Shared these

Heh, the output you gave me did actually not contain info about the range we wanted... It gave us the first 5 ranges and then some crap, and we needed the first 7 ranges. I think the output of that query is a bit random. You can do the following to get the id of the SystemConfigSpan:
```
select * from crdb_internal.ranges where start_pretty like '%SystemConfigSpan%';
```
And then you can take the id from there, if it's not 7 which we already tried, and for a GC Queue run on that range.

But anyway, perhaps in the meantime the cluster has completely recovered? Is the GC Queue now empty (or perhaps became empty after doing the `guest_order@guest_order_guest_id_card_otep_token_idx` thing from above) and all the schema changes are happy and things are good?

**Comment by matbhuvi** [comment]
> You can do the following to get the id of the SystemConfigSpan:

The range is 7 only.

```
root@:26257/postgres> select * from crdb_internal.ranges where start_pretty like '%SystemConfigSpan%';
  range_id | start_key |         start_pretty          | end_key | end_pretty | database_name | table_name | index_name |  replicas   | lease_holder
+----------+-----------+-------------------------------+---------+------------+---------------+------------+------------+-------------+--------------+
         7 | \210      | /Table/SystemConfigSpan/Start | \223    | /Table/11  |               |            |            | {1,2,3,4,5} |            3
(1 row)

Time: 43.018246937s
```

> But anyway, perhaps in the meantime the cluster has completely recovered? Is the GC Queue now empty

Yes. Cluster has recovered. GC queue is empty now. Schema changes are going smooth.

**Comment by andreimatei** [comment]
> The only issue as of now is the problem ranges report. It is still timing out.

I've opened #34311 for this. Let's move the discussion there for the timeout please.
Otherwise, it sounds like we're done here. Thanks for your help!
Feel free to reopen or create new issues for new badness.
